skbio/metadata/_interval.py

```diff
@@ -0,0 +1,334 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from ._intersection import IntervalTree
+
+
+class Interval():
+    '''Stores the position and metadata of an interval.
+
+    Parameters
+    ----------
+    intervals : list of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : list of tuple of bool
+        List of tuples, representing the openness of each interval.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+    interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    '''
+    def __init__(self, intervals=None, boundaries=None,
```

Comment:
@ebolyen and I chatted about the `Interval` constructor and whether it should be private or public. We think it should be public, with the recommended way of adding metadata to an interval being `IntervalMetadata.add` (see my comment [here](https://github.com/biocore/scikit-bio/pull/1356/files#r64837762)). If a user wants, they can accomplish the same thing by using the `Interval` constructor, which is an implicit `add` operation.

We propose the following constructor signature:

```python
def __init__(self, interval_metadata, intervals, boundaries=None, metadata=None):
    ...
```

`interval_metadata` should be required since it isn't a useful operation to construct an unbound `Interval`, and it will be consistent with `IntervalMetadata.add` such that an `Interval` is always (at least initially) bound to `IntervalMetadata`.

`intervals` should be required since there is no way to insert metadata into the tree without intervals and it is the central piece of data here.

Does this API make sense to you guys?